File: python/src/parser.py
"""
Parser Module
-------------

This module takes raw HTML content and extracts links.
It uses BeautifulSoup with lxml for high-performance parsing.
"""
from bs4 import BeautifulSoup
from typing import Set
from .utils import normalize_url, is_same_domain

def extract_links(html_content: str, base_url: str) -> Set[str]:
    """
    Extracts all valid links from the HTML content that belong to the same domain.
    Returns a set of normalized, absolute URLs.
    """
    links = set()
    try:
        soup = BeautifulSoup(html_content, 'lxml')
        for a_tag in soup.find_all('a', href=True):
            href = a_tag['href']
            normalized = normalize_url(base_url, href)
            
            # Filter out non-http/https schemes (e.g., mailto:, tel:, javascript:)
            if not normalized.startswith(('http://', 'https://')):
                continue
                
            if is_same_domain(base_url, normalized):
                links.add(normalized)
    except Exception as e:
        # In a real crawler, we might log this error
        pass
        
    return links

File: golang/internal/parser/parser.go
package parser

import (
	"bytes"
	"strings"

	"github.com/PuerkitoBio/goquery"

	"webcrawler/internal/domain"
)

// Parser extracts links from HTML content.
type Parser struct{}

// NewParser creates a new HTML parser.
func NewParser() *Parser {
	return &Parser{}
}

// ExtractLinks extracts all valid links from HTML content.
// It resolves relative URLs against the base URL.
func (p *Parser) ExtractLinks(baseURL string, htmlContent []byte) ([]string, error) {
	doc, err := goquery.NewDocumentFromReader(bytes.NewReader(htmlContent))
	if err != nil {
		return nil, err
	}

	var links []string
	seen := make(map[string]bool)

	doc.Find("a[href]").Each(func(_ int, s *goquery.Selection) {
		href, exists := s.Attr("href")
		if !exists {
			return
		}

		// Skip empty, javascript, mailto, and fragment-only links
		href = strings.TrimSpace(href)
		if href == "" ||
			strings.HasPrefix(href, "javascript:") ||
			strings.HasPrefix(href, "mailto:") ||
			strings.HasPrefix(href, "tel:") ||
			strings.HasPrefix(href, "data:") ||
			(strings.HasPrefix(href, "#") && len(href) > 1) {
			return
		}

		// Resolve relative URL to absolute
		absoluteURL, err := domain.ResolveURL(baseURL, href)
		if err != nil {
			return
		}

		// Skip non-HTTP URLs
		if !domain.IsValidHTTP(absoluteURL) {
			return
		}

		// Dedupe within this page
		if seen[absoluteURL] {
			return
		}
		seen[absoluteURL] = true

		links = append(links, absoluteURL)
	})

	return links, nil
}

// FilterSameDomain filters links to only include those on the same domain.
func (p *Parser) FilterSameDomain(baseURL string, links []string) []string {
	var filtered []string
	for _, link := range links {
		if domain.IsSameDomain(baseURL, link) {
			filtered = append(filtered, link)
		}
	}
	return filtered
}
